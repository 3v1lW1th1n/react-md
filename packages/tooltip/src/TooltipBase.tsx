import * as React from "react";
import cn from "classnames";
import { CSSTransition } from "react-transition-group";
import {
  CSSTransitionClassNames,
  TransitionTimeout,
} from "@react-md/transition";
import { IWithForwardedRef } from "@react-md/utils";

import { IBaseTooltipProps, TooltipPosition } from "./types.d";

/**
 * All the props that can be applied to the `TooltipBase` component.
 */
export interface ITooltipBaseProps extends IBaseTooltipProps {
  /**
   * This is the position that the tooltip should appear related to its container element as well as
   * updating the animation direction.
   */
  position?: TooltipPosition;

  /**
   * Boolean if the tooltip is visible. This value changing will trigger the different animations.
   */
  visible: boolean;
}

/**
 * An interface containing all the props that have a default value.
 */
export interface ITooltipBaseDefaultProps {
  dense: boolean;
  position: TooltipPosition;
  lineWrap: boolean;
  classNames: CSSTransitionClassNames;
  timeout: TransitionTimeout;
  lazyMount: boolean;
}

export type TooltipBaseWithDefaultProps = ITooltipBaseProps &
  ITooltipBaseDefaultProps &
  IWithForwardedRef<HTMLSpanElement>;

/**
 * This is the base tooltip component that can only be used to render a tooltip with an animation
 * when the visibility changes. If this component is used, you will need to manually add all the
 * event listeners and triggers to change the `visible` prop.
 */
const TooltipBase: React.FunctionComponent<
  ITooltipBaseProps & IWithForwardedRef<HTMLSpanElement>
> = providedProps => {
  const {
    className,
    classNames,
    visible,
    timeout,
    lazyMount,
    dense,
    lineWrap,
    position,
    children,
    forwardedRef,
    onEnter,
    onEntering,
    onEntered,
    onExit,
    onExiting,
    onExited,
    ...props
  } = providedProps as TooltipBaseWithDefaultProps;

  return (
    <CSSTransition
      classNames={classNames}
      in={visible}
      timeout={timeout}
      appear={lazyMount}
      mountOnEnter={lazyMount}
      unmountOnExit={lazyMount}
      onEnter={onEnter}
      onEntering={onEntering}
      onEntered={onEntered}
      onExit={onExit}
      onExiting={onExiting}
      onExited={onExited}
    >
      <span
        {...props}
        ref={forwardedRef}
        role="tooltip"
        className={cn(
          "rmd-tooltip",
          {
            "rmd-tooltip--dense": dense,
            "rmd-tooltip--line-wrap": lineWrap,
            "rmd-tooltip--dense-line-wrap": dense && lineWrap,
          },
          `rmd-tooltip--${position}`,
          className
        )}
      >
        {children}
      </span>
    </CSSTransition>
  );
};

const defaultProps: ITooltipBaseDefaultProps = {
  dense: false,
  position: "below",
  lineWrap: false,
  classNames: {
    enter: "rmd-tooltip--enter",
    enterActive: "rmd-tooltip--visible",
    enterDone: "rmd-tooltip--visible",
    exit: "rmd-tooltip--visible rmd-tooltip--exit",
    exitActive: "rmd-tooltip--exit-active",
  },
  timeout: {
    enter: 200,
    exit: 150,
  },
  lazyMount: true,
};

TooltipBase.defaultProps = defaultProps;

if (process.env.NODE_ENV !== "production") {
  // there's a problem with forwardedRef components that set the `displayName` to `undefined`
  TooltipBase.displayName = "TooltipBase";

  let PropTypes;
  try {
    PropTypes = require("prop-types");
  } catch (e) {}

  if (PropTypes) {
    TooltipBase.propTypes = {
      id: PropTypes.string.isRequired,
      style: PropTypes.object,
      className: PropTypes.string,
      children: PropTypes.node.isRequired,
      dense: PropTypes.bool,
      lineWrap: PropTypes.bool,
      classNames: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.shape({
          appear: PropTypes.string,
          appearActive: PropTypes.string,
          enter: PropTypes.string,
          enterActive: PropTypes.string,
          enterDone: PropTypes.string,
          exit: PropTypes.string,
          exitActive: PropTypes.string,
          exitDone: PropTypes.string,
        }),
      ]),
      timeout: PropTypes.oneOfType([
        PropTypes.number,
        PropTypes.shape({
          enter: PropTypes.number,
          exit: PropTypes.number,
        }),
      ]),
      onEnter: PropTypes.func,
      onEntering: PropTypes.func,
      onEntered: PropTypes.func,
      onExit: PropTypes.func,
      onExiting: PropTypes.func,
      onExited: PropTypes.func,
      lazyMount: PropTypes.bool,
      position: PropTypes.oneOf(["above", "below", "left", "right"]),
      visible: PropTypes.bool.isRequired,
    };
  }
}

export default React.forwardRef<HTMLSpanElement, ITooltipBaseProps>(
  (props, ref) => <TooltipBase {...props} forwardedRef={ref} />
);
