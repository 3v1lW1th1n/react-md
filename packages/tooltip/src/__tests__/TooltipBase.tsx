import * as React from "react";
import { mount } from "enzyme";

import TooltipBase from "../TooltipBase";

describe("<Tooltip />", () => {
  it("should default to setting the tooltip position to below", () => {
    const tooltip = mount(
      <TooltipBase id="tooltip-1" visible={true}>
        This is a tooltip
      </TooltipBase>
    )
      .find("span")
      .first();

    expect(tooltip.hasClass("rmd-tooltip--above")).toBe(false);
    expect(tooltip.hasClass("rmd-tooltip--right")).toBe(false);
    expect(tooltip.hasClass("rmd-tooltip--below")).toBe(true);
    expect(tooltip.hasClass("rmd-tooltip--left")).toBe(false);
  });

  it("should correctly apply the position suffix", () => {
    const wrapper = mount(
      <TooltipBase id="tooltip-1" position="above" visible={true}>
        This is a tooltip
      </TooltipBase>
    );

    let tooltip = wrapper.find("span").first();
    expect(tooltip.hasClass("rmd-tooltip--above")).toBe(true);
    expect(tooltip.hasClass("rmd-tooltip--right")).toBe(false);
    expect(tooltip.hasClass("rmd-tooltip--below")).toBe(false);
    expect(tooltip.hasClass("rmd-tooltip--left")).toBe(false);

    wrapper.setProps({ position: "right" });
    tooltip = wrapper.find("span").first();
    expect(tooltip.hasClass("rmd-tooltip--above")).toBe(false);
    expect(tooltip.hasClass("rmd-tooltip--right")).toBe(true);
    expect(tooltip.hasClass("rmd-tooltip--below")).toBe(false);
    expect(tooltip.hasClass("rmd-tooltip--left")).toBe(false);

    wrapper.setProps({ position: "below" });
    tooltip = wrapper.find("span").first();
    expect(tooltip.hasClass("rmd-tooltip--above")).toBe(false);
    expect(tooltip.hasClass("rmd-tooltip--right")).toBe(false);
    expect(tooltip.hasClass("rmd-tooltip--below")).toBe(true);
    expect(tooltip.hasClass("rmd-tooltip--left")).toBe(false);

    wrapper.setProps({ position: "left" });
    tooltip = wrapper.find("span").first();
    expect(tooltip.hasClass("rmd-tooltip--above")).toBe(false);
    expect(tooltip.hasClass("rmd-tooltip--right")).toBe(false);
    expect(tooltip.hasClass("rmd-tooltip--below")).toBe(false);
    expect(tooltip.hasClass("rmd-tooltip--left")).toBe(true);
  });

  it("should render correctly", () => {
    const tooltip = mount(
      <TooltipBase id="tooltip-1" visible={false}>
        This is some content
      </TooltipBase>
    );
    expect(tooltip.render()).toMatchSnapshot();

    tooltip.setProps({ visible: true });
    expect(tooltip.render()).toMatchSnapshot();

    tooltip.setProps({ dense: true });
    expect(tooltip.render()).toMatchSnapshot();
  });
});
