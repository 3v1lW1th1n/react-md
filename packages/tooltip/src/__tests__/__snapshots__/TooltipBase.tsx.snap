// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tooltip /> should render correctly 1`] = `null`;

exports[`<Tooltip /> should render correctly 2`] = `
<span
  class="rmd-tooltip rmd-tooltip--below rmd-tooltip--enter rmd-tooltip--visible"
  id="tooltip-1"
  role="tooltip"
>
  This is some content
</span>
`;

exports[`<Tooltip /> should render correctly 3`] = `
<span
  class="rmd-tooltip rmd-tooltip--dense rmd-tooltip--below"
  id="tooltip-1"
  role="tooltip"
>
  This is some content
</span>
`;
